### PR TITLE
Adding PeripheralRingMeshGenerator

### LINKED ISSUE #19978: Adding Peripheral Ring Region Meshing Capability to Reactor Module
## Reason
The `Reactor` module of MOOSE has already been equipped with a variety of mesh generators to mesh complex geometries involved in reactor cores. However, the current implementation does not include the capability of meshing the peripheral ring regions of reactors with cylindrical shape. This issue aims to add this new feature to the `Reactor` module.

## Design
A new mesh generator named `PeripheralRingMeshGenerator` will be added, which is capable of meshing the peripheral ring region of a reactor given the internal mesh with a specified external boundary.

## Impact
With this new feature described above, MOOSE users will be able to mesh the entire cylindrical reactor cores with the peripheral ring regions.

### PR DESCRIPTION
**Pull request by miaoyinb**
closes #19978

### CODE REVIEW

**Comment by miaoyinb** [comment]
@eshemon @shikhar413 FYI

**Comment by miaoyinb** [comment]
Happy New Year! @loganharbour @GiudGiud
This PR is to add peripheral ring meshing capability for a reactor core to the `Reactor` module. Thank you in advance for your review!

**Comment by YaqiWang** [comment]
If we have poly2tri, do we still need this capability? The mesh generated in this does not look very nice.

**Comment by eshemon** [comment]
@YaqiWang yes, this is complementary to the other capability. Actually the mesh looks much nicer than what you will get with poly2tri for a heterogeneous assembly. There is no magic bullet. One will work better than the other depending on your core mesh.

**Comment by YaqiWang** [comment]
Oh? I thought poly2tri should have a control parameter like other meshers that can control the maximum aspect ratio of elements it generates. This is not the case?

**Comment by aaronoaks** [comment]
Poly2tri is a fairly simple triangulation library. It is mainly focused on generating a triangulation of a given region but without much regard for the quality of the triangles. It has the option to take in Steiner points that can be used to improve the quality of the triangles, and I added some functionality to the related mesh generator to use this option to improve the mesh some, but it is not an automated refinement algorithm.

I believe there are things in the works under MOOSE's scope to improve triangulation capabilities in the longer term, @eshemon may be able to go into detail.

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    requirement ='The system shall add a peripheral ring to to the input mesh without preserving volumes due to polygonization.'
+    recover = false
+    mesh_mode = 'REPLICATED'
+  []
+  [core_peripheral_ring_preserved]
+    type = 'Exodiff'
+    input = 'core_ring.i'
+    exodiff = 'core_ring_preserved_in.e'
+    cli_args = '--mesh-only core_ring_preserved_in.e'
+    requirement ='The system shall add a peripheral ring to to the input mesh with preserving volumes due to polygonization.'
```
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:19]:
```suggestion
    requirement ='The system shall be able to add a peripheral ring to to the input mesh and preserve volumes even with  polygonization.'
```

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    recover = false
+    mesh_mode = 'REPLICATED'
+  []
+  [err_invalid_boundary]
+    type = 'RunException'
+    input = 'core_ring.i'
+    cli_args = 'Mesh/pr/input_mesh_external_boundary=1
+                --mesh-only'
+    expect_err = 'This mesh generator does not work for the provided boundary.'
+    requirement ='The system shall threw an error if the given boundary of the input mesh is not valid.'
```
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:29]:
```suggestion
    requirement ='The system shall throw an error if the given boundary of the input mesh is not valid.'
```

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    mesh_mode = 'REPLICATED'
+  []
+  [err_small_radius]
+    type = 'RunException'
+    input = 'core_ring.i'
+    cli_args = 'Mesh/pr/peripheral_ring_radius=20
+                --mesh-only'
+    expect_err = 'The peripheral ring to be generated must be large enough to cover the entire input mesh.'
+    requirement ='The system shall threw an error if the given peripheral ring radius is not enough to cover the input mesh.'
```
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:38]:
```suggestion
    requirement ='The system shall throw an error if the given peripheral ring radius is not enough to cover the input mesh.'
```

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    mesh_mode = 'REPLICATED'
+  []
+  [err_internal_boundary]
+    type = 'RunException'
+    input = 'core_ring.i'
+    cli_args = 'Mesh/fmg/file=input_mesh_err.e
+                Mesh/pr/input_mesh_external_boundary=30500
+                --mesh-only'
+    expect_err = 'The boundary provided is not an external boundary.'
+    requirement ='The system shall threw an error if the given external boundary is actually an internal boundary of the input mesh.'
```
Comment (GiudGiud) [inline modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:48]:
```suggestion
    requirement ='The system shall throw an error if the given external boundary is actually an internal boundary of the input mesh.'
```

File: modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests (hunk 3 above)
Comment (GiudGiud) [reply modules/reactor/test/tests/meshgenerators/peripheral_ring_mesh_generator/tests:38]: the peripheral mesh radius is not 'covering' the input mesh. Let's rephrase like "circle the mesh" or something

**Review by GiudGiud — COMMENTED** [review]
will review more later

File: modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  /// Subdomain ID of the added peripheral region
+  const subdomain_id_type _peripheral_ring_block_id;
+  /// Subdomain name of the added peripheral region
+  const SubdomainName _peripheral_ring_block_name;
+  /// ID of the new external boundary
+  const boundary_id_type _external_boundary_id;
+  /// Name of the new external boundary
+  const std::string _external_boundary_name;
+  /// Pointer to the input mesh
+  std::unique_ptr<MeshBase> * _mesh_ptrs;
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:47]: will use reference instead

File: modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  /// Subdomain ID of the added peripheral region
+  const subdomain_id_type _peripheral_ring_block_id;
+  /// Subdomain name of the added peripheral region
+  const SubdomainName _peripheral_ring_block_name;
+  /// ID of the new external boundary
+  const boundary_id_type _external_boundary_id;
+  /// Name of the new external boundary
+  const BoundaryName _external_boundary_name;
+  /// Reference to input mesh pointer
+  std::unique_ptr<MeshBase> & _mesh;
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:48]: you only use this attribute in one place. I'd remove it

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+  const Point origin_pt = MooseMeshUtils::meshCentroidCalculator(*input_mesh);
+  const Real origin_x = origin_pt(0);
+  const Real origin_y = origin_pt(1);
+  // Vessel for containing maximum radius of the boundary nodes
+  Real max_input_mesh_node_radius;
+
+  if (!isBoundaryValid(
+          *input_mesh, max_input_mesh_node_radius, origin_pt, _input_mesh_external_bid))
+    paramError("input_mesh_external_boundary",
+               "This mesh generator does not work for the provided boundary.");
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:88]: let's try to add a list of possible reasons there so a user has somewhere to go from there

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+bool
+PeripheralRingMeshGenerator::isExternalBoundary(ReplicatedMesh & mesh,
+                                                const boundary_id_type bid) const
+{
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto side_list = boundary_info.build_side_list();
+  for (unsigned int i = 0; i < side_list.size(); i++)
+  {
+    if (std::get<2>(side_list[i]) == bid)
+      if (mesh.elem_ptr(std::get<0>(side_list[i]))->neighbor_ptr(std::get<1>(side_list[i])) !=
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:332]: we ll have to make sure the mesh is prepared for use or at least that find_neighbors has been called

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+  for (unsigned int i = 0; i < node_list.size(); ++i)
+    if (std::get<1>(node_list[i]) == _input_mesh_external_bid)
+    {
+      input_ext_node_num++;
+      // Define nodes on the inner and outer boundaries of the peripheral region.
+      input_ext_bd_pts.push_back(input_mesh->point(std::get<0>(node_list[i])));
+      tmp_azi = atan2(input_ext_bd_pts.back()(1) - origin_y, input_ext_bd_pts.back()(0) - origin_x);
+      output_ext_bd_pts.push_back(Point(_peripheral_ring_radius * std::cos(tmp_azi),
+                                        _peripheral_ring_radius * std::sin(tmp_azi),
+                                        0.0));
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:129]: 0 might not be the z-coordinate of the input mesh

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+      // a tuple using azimuthal angle as the key to faciliate sorting
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:130]:
```suggestion
      // a tuple using azimuthal angle as the key to facilitate sorting

```

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 11]
```diff
@@ -0,0 +1,337 @@
...
+      azi_points.push_back(
+          std::make_tuple(tmp_azi, input_ext_bd_pts.back(), output_ext_bd_pts.back()));
+      // a simple vector of azimuthal angles for radius correction purpose (in degree)
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:133]:
```suggestion
      // a simple vector of azimuthal angles for radius correction purposes (in degree)

```

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 12]
```diff
@@ -0,0 +1,337 @@
...
+  std::sort(azi_points.begin(), azi_points.end());
+  std::sort(azi_array.begin(), azi_array.end());
+  const Real correction_factor = _preserve_volumes ? radiusCorrectionFactor(azi_array) : 1.0;
+  for (unsigned int i = 0; i < input_ext_node_num; ++i)
+  {
+    // Inner boundary nodes of the peripheral region
+    points_array[0][i] = std::get<1>(azi_points[i]);
+    // Outer boundary nodes of the peripheral region
+    points_array[_peripheral_layer_num][i] = std::get<2>(azi_points[i]) * correction_factor;
+    // Use interpolation to get intermediated layers nodes
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:145]:
```suggestion
    // Use interpolation to get intermediate layers nodes

```

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 13]
```diff
@@ -0,0 +1,337 @@
...
+      elem_Quad4->set_node(3) = nodes[node_id_array[i][(j + 1) % input_ext_node_num]];
+      elem_Quad4->subdomain_id() = _peripheral_ring_block_id;
+
+      if (i == 0)
+        boundary_info.add_side(elem_Quad4, 3, OUTER_SIDESET_ID_ALT);
+      if (i == _peripheral_layer_num - 1)
+        boundary_info.add_side(elem_Quad4, 1, OUTER_SIDESET_ID);
+    }
+
+  // This would make sure that the boundary OUTER_SIDESET_ID is deleted after stiching.
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:181]:
```suggestion
  // This would make sure that the boundary OUTER_SIDESET_ID is deleted after stitching.

```

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+      std::reverse(boundary_orderred_node_list.begin(), boundary_orderred_node_list.end());
+      // As this iteration is wasted, set the iterator backward
+      i--;
+    }
+  }
+  // If the code ever gets here, boundary_node_assm is empty.
+  // If the isFlipped == true, the boundary is not a loop.
+  // This is not done inside the loop just for some potential applications in the future.
+  if (isFlipped)
+    return false;
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:289]: why is it not a loop if it's flipped here? What's the point of retrying the with the reverse boundary nodes then?

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+    {
+      orderred_node_azi_list.push_back(
+          (*mesh.node_ptr(boundary_orderred_node_list[i]) - origin_pt)
+              .cross(*mesh.node_ptr(boundary_orderred_node_list[i + 1]) - origin_pt)(2));
+      // Use this opportunity to calculate maximum radius
+      max_node_radius = std::max(
+          (*mesh.node_ptr(boundary_orderred_node_list[i]) - origin_pt).norm(), max_node_radius);
+    }
+    std::sort(orderred_node_azi_list.begin(), orderred_node_azi_list.end());
+    if (orderred_node_azi_list.front() * orderred_node_azi_list.back() < 0.0)
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:307]: so in the case that two nodes and the origin are aligned, we get 0 there right? We cant really make the mesh then right? so maybe let's have a tolerance on that 0 ?

File: modules/reactor/test/tests/meshgenerators/patterned_hex_mesh_generator/positions_meta.data [hunk 16]
```diff
@@ -1 +1 @@
-52.1836 -30.1282 0.0
+43.3013 -25 0.0
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/test/tests/meshgenerators/patterned_hex_mesh_generator/positions_meta.data:1]: The previous code use the average of all external boundary nodes as the origin position, which works well for symmetric reactor cores. Unfortunately this simple example mesh is not symmetric so that the origin is not the mesh centroid (it does not cause any other issue though). With the new mesh centroid calculator, this is fixed.

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C (hunk 14 above)
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:289]: A closed loop should have no ends so that flipping is not needed to traverse all the nodes. For an open curve, if we do not start from one end, then we will need to "flip" once to traverse all the nodes. I try to traverse all the nodes after flipping for some possible future applications.

File: modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h (hunk 6 above)
Comment (miaoyinb) [reply modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:48]: I tried not to use it as a member variable. It seems that MOOSE is confused with the mesh generator dependencies and throws an error. Any suggestion?

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C (hunk 9 above)
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:129]: Will allow nonzero z coordinate meshes.

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C (hunk 15 above)
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:307]: If two nodes and the origin are aligned, it should be sure that there are positive AND negative elements in `orderred_node_azi_list` and this one would be < 0.0 (because the origin is the centroid, there should be nodes on both sides of that line.)

**Comment by miaoyinb** [comment]
@loganharbour @GiudGiud Please let us know if any additional revision is needed. We would appreciate it a lot if this PR could be  merged soon. We have some interested users looking forward to this new feature. Thanks.

File: modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:48]:
```suggestion
  std::unique_ptr<MeshBase> & _input;
```
it's the name used everywhere else. 
If you could rename it in AzimuthalBlockSplitGenerator.h would be good too

**Review by GiudGiud — CHANGES_REQUESTED** [review]
going to be good for me after we've discussed/fixed these last few things

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+        boundary_info.add_side(elem_Quad4, 3, OUTER_SIDESET_ID_ALT);
+      if (i == _peripheral_layer_num - 1)
+        boundary_info.add_side(elem_Quad4, 1, OUTER_SIDESET_ID);
+    }
+
+  // This would make sure that the boundary OUTER_SIDESET_ID is deleted after stitching.
+  if (_input_mesh_external_bid != OUTER_SIDESET_ID)
+    MooseMesh::changeBoundaryId(*input_mesh, _input_mesh_external_bid, OUTER_SIDESET_ID, true);
+  mesh->prepare_for_use();
+  mesh->stitch_meshes(*input_mesh, OUTER_SIDESET_ID_ALT, _input_mesh_external_bid, TOLERANCE, true);
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:212]: I agree that this is kinda expensive. The good news is it's still 2D. We tried some actual reactor core meshes and it seems to work well.

File: modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h (hunk 17 above)
Comment (miaoyinb) [reply modules/reactor/include/meshgenerators/PeripheralRingMeshGenerator.h:48]: will do

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+                                : std::string()),
+    _mesh(getMeshByName(_input_name))
+{
+}
+
+std::unique_ptr<MeshBase>
+PeripheralRingMeshGenerator::generate()
+{
+  // Need ReplicatedMesh for stitching
+  auto input_mesh = dynamic_cast<ReplicatedMesh *>(_mesh.get());
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:77]: will do

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C (hunk 18 above)
[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:212]: Ok I just tried with 20k boundary elements and it was like 8s so that's reasonable

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 20]
```diff
@@ -0,0 +1,379 @@
...
+
+  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> side_list =
+      input_mesh->get_boundary_info().build_side_list();
+  input_mesh->get_boundary_info().build_node_list_from_side_list();
+  std::vector<std::tuple<dof_id_type, boundary_id_type>> node_list =
+      input_mesh->get_boundary_info().build_node_list();
+
+  // 2D vector containing all the node positions of the peripheral region
+  std::vector<std::vector<Point>> points_array(_peripheral_layer_num + 1,
+                                               std::vector<Point>(node_list.size()));
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:139]: It is the correct size. This variable contains all the Points needed for the peripheral region. As the peripheral region here has a layered structure, the number of nodes in each layer is the same as the number of nodes on the given boundary.

[the remark this answers is not in the document]
Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:142]: but node_list is not the size of the given boundary?
```
  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> side_list =
      input_mesh->get_boundary_info().build_side_list();
  input_mesh->get_boundary_info().build_node_list_from_side_list();
  std::vector<std::tuple<dof_id_type, boundary_id_type>> node_list =
      input_mesh->get_boundary_info().build_node_list();
```
I dont see where a specific sideset is being targetted?

[the remark this answers is not in the document]
Comment (miaoyinb) [reply modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:142]: You are right. I am using another `node_list` variable in another project and I was confused when writing these lines, wasting some memory. Will fix this.

**Comment by miaoyinb** [comment]
@GiudGiud Any idea why civet is pending for so long? Thanks.

**Comment by miaoyinb** [comment]
> > @GiudGiud Any idea why civet is pending for so long? Thanks.
> 
> Services are down.

thanks.

**Comment by miaoyinb** [comment]
> Job [Precheck](https://civet.inl.gov/job/935756/) on [f50bdb3](https://github.com/idaholab/moose/commit/f50bdb3b39b2427fcce991fe4fc3affa0e1c7257) wanted to post the following:
> 
> Your code requires style changes.
> 
> A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19979/style.patch) You can directly apply the patch by running, in the top level of your repository:
> 
> `curl -s https://mooseframework.inl.gov/docs/PRs/19979/style.patch | git apply -v`
> 
> Alternatively, with your repository up to date and in the top level of your repository:
> 
> `git clang-format 6c9077993af12d6b56619cb67309a89f34871918`

I must be using the wrong version of clang-format. Any suggestion?

**Comment by loganharbour** [comment]
Just run:

```
curl -s https://mooseframework.inl.gov/docs/PRs/19979/style.patch | git apply -v
```

as it suggests in the top level of the repo and commit those changes.

We're working on unifying the version of clang-format on the build boxes vs the one we provide with conda.

**Review by GiudGiud — DISMISSED** [review]

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 21]
```diff
@@ -0,0 +1,382 @@
...
+                                             const Point origin_pt,
+                                             const boundary_id_type bid) const
+{
+  max_node_radius = 0.0;
+  invalid_type = 0;
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto side_list_tmp = boundary_info.build_side_list();
+  unsigned int elem_counter = 0;
+  std::vector<std::pair<dof_id_type, dof_id_type>> boundary_node_assm;
+  std::vector<dof_id_type> boundary_orderred_node_list;
```
Comment (roystgnr) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:250]: ordered, one r?

File: modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+  bool isFlipped = false;
+  for (unsigned int i = 0; i < side_list_tmp.size(); i++)
+  {
+    if (std::get<2>(side_list_tmp[i]) == bid)
+    {
+      elem_counter++;
+      // store two nodes of each side
```
Comment (roystgnr) [inline modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C:257]: Do we have tests + moose errors anywhere here for the `elem->dim() != 2` case?  It's obvious from the docs that only 2D meshes should be supported, but if someone messes up and tries to use this with a 3D or mixed-dimension mesh anyway we'd like them to get a sane error message out.

**Review by roystgnr — CHANGES_REQUESTED** [review]

File: framework/src/utils/MooseMeshUtils.C [hunk 23]
```diff
@@ -165,4 +165,21 @@ getSubdomainID(const SubdomainName & subdomain_name, const MeshBase & mesh)
...
+  // origin.
+  for (const auto & elem : as_range(mesh.elements_begin(), mesh.elements_end()))
+  {
+    Real elem_vol = elem->volume();
+    origin_pt += (elem->true_centroid()) * elem_vol;
+    vol_tmp += elem_vol;
+  }
+  origin_pt /= vol_tmp;
+  return origin_pt;
+}
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply framework/src/utils/MooseMeshUtils.C:184]: Yes, it was a class member function of `PolygonMeshGeneratorBase` in the original version for origin calculation. I was suggested to move it into `MooseMeshUtils` as it could be a general tool.

[the remark this answers is not in the document]
Comment (miaoyinb) [reply framework/src/utils/MooseMeshUtils.C:186]: I corrected this and generalized it as you suggested. I have not worked on `libMesh::Parallel::Communicator` much. Hopefully it's done correctly. 😅

**Comment by miaoyinb** [comment]
@loganharbour @roystgnr Please take a look to see if anything else needs to be addressed, thanks.

File: framework/src/utils/MooseMeshUtils.C [hunk 24]
```diff
@@ -167,19 +166,22 @@ getSubdomainID(const SubdomainName & subdomain_name, const MeshBase & mesh)
...
+       as_range(mesh.active_local_elements_begin(), mesh.active_local_elements_end()))
   {
     Real elem_vol = elem->volume();
-    origin_pt += (elem->true_centroid()) * elem_vol;
+    centroid_pt += (elem->true_centroid()) * elem_vol;
     vol_tmp += elem_vol;
   }
-  origin_pt /= vol_tmp;
-  return origin_pt;
+  mesh.comm().sum(centroid_pt(0));
```
[the remark this answers is not in the document]
Comment (miaoyinb) [reply framework/src/utils/MooseMeshUtils.C:181]: Yes, I got a compiler error. Will try your suggestion.

File: framework/src/utils/MooseMeshUtils.C [hunk 25]
```diff
@@ -12,6 +12,7 @@
 
 #include "libmesh/elem.h"
 #include "libmesh/boundary_info.h"
+#include "libmesh/replicated_mesh.h"
```
[the remark this answers is not in the document]
Comment (roystgnr) [reply framework/src/utils/MooseMeshUtils.C:15]: Ah - my apologies - I was looking at the commits here one at a time and didn't realize that this diff was restoring something that a previous commit had removed.

I do think it's not needed; looking over the code suggests that some idiot introduced it for no good reason ... but don't delete it, leave it in.  `git log` reveals that the idiot was *me*, and the log message gives a plausible reason; just in case I was correct and the reason is still subtly valid let's not take the risk of regressing something here.

**Review by roystgnr — DISMISSED** [review]
This would be better with the centroid Point sum done in one line, but that's a nitpick; I'm happy with it now either way.

File: framework/src/utils/MooseMeshUtils.C (hunk 24 above)
[the remark this answers is not in the document]
Comment (miaoyinb) [reply framework/src/utils/MooseMeshUtils.C]: Works

**Review by loganharbour — DISMISSED** [review]

File: modules/reactor/test/tests/functions/multi_control_drum_function/positions_meta.data [hunk 26]
```diff
@@ -1,6 +1,6 @@
 6.06218 3.5 0.0
--5.78862e-15 7 0.0
```
Comment (roystgnr) [inline modules/reactor/test/tests/functions/multi_control_drum_function/positions_meta.data]: Why is this file changed here?

**Comment by eshemon** [comment]
@cticenhour this MR looks almost done, and it's similar in functionality to the one Roy and Aaron just finished (featured in Jan newsletter). Just a heads up in case there is time to advertise this one in there too

**Comment by cticenhour** [comment]
@eshemon Happy to have a new entry if this will be going in before Feb 1! If you don't want to add it directly to this PR in a new commit, you can send me the text on slack (and a reference to an image if you'd like one to be featured), and I can add it in a small follow-up.

**Comment by roystgnr** [comment]
> if this will be going in before Feb 1!

It should go in this afternoon.

My last comment above is out of concern that `sum(Point)` might be behaving differently in some way from individual component sums - the two should have identical behavior, not just "identical to within a few ULPs" behavior.  But even if there's something I need to be looking at in TIMPI, that shouldn't delay merging here as soon as CI is done again.

File: modules/reactor/test/tests/functions/multi_control_drum_function/positions_meta.data (hunk 26 above)
Comment (miaoyinb) [reply modules/reactor/test/tests/functions/multi_control_drum_function/positions_meta.data]: This is a "zero" calculated by the new `sum(Point)`.

Comment (roystgnr) [reply modules/reactor/test/tests/functions/multi_control_drum_function/positions_meta.data]: Thanks, I see.

But now that I think about the problem: why bother changing it in the repo?  MPI_Reduce operations don't guarantee any particular operation ordering, and floating point addition isn't associative so you can't expect the results to go unchanged if you reorder operations, so even if we hadn't changed which `sum()` to use, we'd still expect the final result to change by some multiple of machine precision even if we re-run it on the same MPI implementation on the same machine on the same number of (more than two) processors.  (and we'd expect it to change more if any of those things change).

So I'm worried: you bothering to commit this file suggests that O(1e-14) changes here are really important ... are they?  If that's the case then we must be doing something wrong elsewhere, because O(1e-14) changes here are unavoidable.

Comment (miaoyinb) [reply modules/reactor/test/tests/functions/multi_control_drum_function/positions_meta.data]: I agree with you. I will remove this file and add into .gitignore. It is generated for users but not useful for this regression test.

**Review by roystgnr — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Comment by miaoyinb** [comment]
FYI @eshemon @aprilnovak
